sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/SdkTracerProviderBuilder.scala

```diff
@@ -0,0 +1,155 @@
+import org.typelevel.otel4s.sdk.context.propagation.TextMapPropagator
+import org.typelevel.otel4s.sdk.trace.samplers.Sampler
+
+/** Builder for [[SdkTracerProvider]].
+  */
+sealed trait SdkTracerProviderBuilder[F[_]] {
+
+  /** Sets an [[IdGenerator]].
+    *
+    * [[IdGenerator]] will be used each time a
+    * [[org.typelevel.otel4s.trace.Span Span]] is started.
+    *
+    * ''Note'': the id generator must be thread-safe and return immediately (no
+    * remote calls, as contention free as possible).
+    */
+  def setIdGenerator(idGenerator: IdGenerator[F]): SdkTracerProviderBuilder[F]
+
+  /** Sets a [[org.typelevel.otel4s.sdk.Resource]] to be attached to all spans
+    * created by [[org.typelevel.otel4s.trace.Tracer Tracer]].
+    */
+  def setResource(resource: InstrumentResource): SdkTracerProviderBuilder[F]
+
+  /** Merges a [[org.typelevel.otel4s.sdk.Resource]] with the current one.
+    */
+  def addResource(resource: InstrumentResource): SdkTracerProviderBuilder[F]
+
+  /** Sets an initial [[SpanLimits]] that should be used with this SDK.
+    *
+    * The limits will be used for every
+    * [[org.typelevel.otel4s.trace.Span Span]].
+    */
+  def setSpanLimits(limits: SpanLimits): SdkTracerProviderBuilder[F]
+
+  /** Sets a [[org.typelevel.otel4s.sdk.trace.samplers.Sampler Sampler]] to use
+    * for sampling traces.
+    *
+    * Sampler will be called each time a
+    * [[org.typelevel.otel4s.trace.Span Span]] is started.
+    *
+    * ''Note:'' the sampler must be thread-safe and return immediately (no
+    * remote calls, as contention free as possible).
+    */
+  def setSampler(sampler: Sampler): SdkTracerProviderBuilder[F]
+
+  /** Adds a [[TextMapPropagator]]s to use for the context propagation.
+    */
+  def addTextMapPropagators(
+      propagators: TextMapPropagator*
+  ): SdkTracerProviderBuilder[F]
+
+  /** Adds a [[SpanProcessor]] to the span processing pipeline that will be
+    * built.
+    *
+    * [[SpanProcessor]] will be called each time a
+    * [[org.typelevel.otel4s.trace.Span Span]] is started or ended.
+    *
+    * ''Note:'' the span processor must be thread-safe and return immediately
+    * (no remote calls, as contention free as possible).
+    */
+  def addSpanProcessor(
+      processor: SpanProcessor[F]
+  ): SdkTracerProviderBuilder[F]
+
+  /** Creates a new [[SdkTracerProvider]] with configuration of this builder.
+    */
+  def build: F[SdkTracerProvider[F]]
+}
+
+object SdkTracerProviderBuilder {
+
+  /** Creates a new [[SdkTracerProviderBuilder]] with default configuration.
+    */
+  def default[F[_]: Async: LocalContext]: SdkTracerProviderBuilder[F] =
+    new Builder[F](
+      idGeneratorF = IdGenerator.default[F],
+      resource = InstrumentResource.Default,
+      spanLimits = SpanLimits.Default,
+      sampler = Sampler.parentBased(Sampler.AlwaysOn),
+      propagators = Nil,
+      spanProcessors = Nil
+    )
+
+  private final case class Builder[F[_]: Async: LocalContext](
```

Comment:
Why do we need `Async` constraint here?

Reply:
https://github.com/typelevel/otel4s/pull/325/files/e0242f732c6aba12db838b1fb96444e7fab7e948#r1342133565 - here is the explanation